docfx~/animator-services.md

```diff
@@ -0,0 +1,301 @@
+
+// Add the clip to a state
+var state = layer.StateMachine.AddState("MyState", motion: virtualClip);
+```
+
+## Working with VirtualControllerContext
+
+The `VirtualControllerContext` provides access to platform-specific animation layers, particularly useful for VRChat avatar manipulation.
+
+### Accessing VRChat Animation Layers
+
+```csharp
+var controllerContext = animatorServices.ControllerContext;
+
+// Access specific VRChat layers (requires VRCSDK3)
+#if NDMF_VRCSDK3_AVATARS
+using VRC.SDK3.Avatars.Components;
+
+// Get the FX layer controller
+if (controllerContext.Controllers.TryGetValue(VRCAvatarDescriptor.AnimLayerType.FX, out var fxController))
+{
+    // Add a new layer to FX
+    var newLayer = fxController.AddLayer(LayerPriority.Default, "MyCustomLayer");
+    
+    // Create states and transitions
+    var idleState = newLayer.StateMachine.AddState("Idle");
+    var activeState = newLayer.StateMachine.AddState("Active", motion: myVirtualClip);
+    
+    // Add transition with conditions
+    var transition = idleState.AddTransition(activeState);
+    transition.AddCondition(AnimatorConditionMode.If, 0, "MyParameter");
+}
+
+// Access other layers
+var baseController = controllerContext.Controllers[VRCAvatarDescriptor.AnimLayerType.Base];
+var additiveController = controllerContext.Controllers[VRCAvatarDescriptor.AnimLayerType.Additive];
+var gestureController = controllerContext.Controllers[VRCAvatarDescriptor.AnimLayerType.Gesture];
+var actionController = controllerContext.Controllers[VRCAvatarDescriptor.AnimLayerType.Action];
+#endif
+```
+
+### Adding and Removing Layers
+
+```csharp
+// Add a new layer
+var newLayer = virtualController.AddLayer(LayerPriority.Default, "MyLayer");
+newLayer.Weight = 1.0f;
+newLayer.BlendingMode = AnimatorLayerBlendingMode.Override;
+
+// Remove a layer
+virtualController.RemoveLayer("UnwantedLayer");
+
+// Modify existing layer
+var existingLayer = virtualController.Layers.FirstOrDefault(l => l.Name == "ExistingLayer");
+if (existingLayer != null)
+{
+    existingLayer.Weight = 0.5f;
+    existingLayer.BlendingMode = AnimatorLayerBlendingMode.Additive;
+}
+```
+
+## Using AnimationIndex and ObjectPathRemapper
+
+### Path Remapping and Object Tracking
+
+The `ObjectPathRemapper` is crucial for handling object hierarchy changes during the build process:
+
+```csharp
+var pathRemapper = animatorServices.ObjectPathRemapper;
+
+// Get the virtual path for a GameObject
+string virtualPath = pathRemapper.GetVirtualPathForObject(myGameObject);
+
+// Record a new object tree (for prefab instantiation)
+GameObject newPrefab = PrefabUtility.InstantiatePrefab(somePrefab) as GameObject;
+pathRemapper.RecordObjectTree(newPrefab.transform);
+
+// Replace an object before removing it
+pathRemapper.ReplaceObject(oldObject, newObject);
+
+// Get final path mappings (typically done automatically on deactivation)
+var pathMappings = pathRemapper.GetVirtualToRealPathMap();
```

Comment:
This is done automatically when the AnimatorServiceContext is closed, and so there is normally no need to invoke this.